### PR TITLE
use std::lerp, constexpr hex.h

### PR DESCRIPTION
**Pull request by myrrc**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Recreation of https://github.com/ClickHouse/ClickHouse/pull/29685

### CODE REVIEW

**Comment by tavplubix** [comment]
It's great that now you have time to finish this PR.
Btw, do you have time to fix https://github.com/ClickHouse/ClickHouse/issues/31768?

**Comment by myrrc** [comment]
> It's great that now you have time to finish this PR. Btw, do you have time to fix #31768 and #43928?

Yes, I'll fix them soon

**Comment by myrrc** [comment]
@tavplubix Maybe you could review?

**Comment by alexey-milovidov** [comment]
The code LGTM. Although I don't know why we need constexpr hex.
Let's fix the build and merge.

**Comment by myrrc** [comment]
> The code LGTM. Although I don't know why we need constexpr hex. Let's fix the build and merge.

There is a (quite negligible) effect on compilation times, however, the main reason for this PR is logical decoupling. `hex.h` and `interpolateLinear.h` do not depend on CH source code thus they were moved to `base/`.

File: base/base/hex.h [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    {
+        TUInt value;
+        UInt8 uint8[sizeof(TUInt)];
+    };
+
+    value = uint_;
+
+    /// Use little endian
+    for (size_t i = 0; i < sizeof(TUInt); ++i)
+        memcpy(out + i * 2, &table[static_cast<size_t>(uint8[sizeof(TUInt) - 1 - i]) * 2], 2);
```
Comment (tavplubix) [inline base/base/hex.h:121]: Please note that ClickHouse can be used on big endian platforms, such as s390x:
https://github.com/ClickHouse/ClickHouse/pulls?q=is%3Apr+big+endian+is%3Aclosed

File: src/Storages/MergeTree/SimpleMergeSelector.cpp [hunk 2]
```diff
@@ -28,7 +28,7 @@ struct Estimator
         {
             double difference = std::abs(log2(static_cast<double>(sum_size) / size_prev_at_left));
             if (difference < settings.heuristic_to_align_parts_max_absolute_difference_in_powers_of_two)
-                current_score *= interpolateLinear(settings.heuristic_to_align_parts_max_score_adjustment, 1,
+                current_score *= std::lerp(settings.heuristic_to_align_parts_max_score_adjustment, 1,
```
Comment (tavplubix) [inline src/Storages/MergeTree/SimpleMergeSelector.cpp:31]: `interpolateLinear` was much more readable than `std::lerp` but okay...

**Review by tavplubix — APPROVED** [review]
Everything else LGTM

File: base/base/hex.h (hunk 1 above)
Comment (myrrc) [reply base/base/hex.h]: Sorry, missed that. Fixed

File: src/Storages/MergeTree/SimpleMergeSelector.cpp (hunk 2 above)
Comment (myrrc) [reply src/Storages/MergeTree/SimpleMergeSelector.cpp:31]: I believe it's better to use std when possible

**Comment by alexey-milovidov** [comment]
> There is a (quite negligible) effect on compilation times, however, the main reason for this PR is logical decoupling.

Just in case: we don't need it at all. I also thought of removing `base` (by merging it into `dbms`) for simplification.

**Pull request merged**